src/libexpr/primops.cc

```diff
@@ -4082,37 +4082,130 @@ void EvalState::createBaseEnv()
 
     /* `builtins' must be first! */
     v.mkAttrs(buildBindings(128).finish());
-    addConstant("builtins", v);
+    addConstant("builtins", v, {
+        .type = "attribute set",
+        .doc = R"(
+          Contains all the [built-in functions](./builtins.md) and values, in order to avoid polluting the global scope.
+
+          Since built-in functions were added over time, [testing for attributes](./operators.md#has-attribute) in `builtins` can be used for graceful fallback on older Nix installations:
+
+          ```nix
+          if builtins ? getEnv then builtins.getEnv "PATH" else ""
+          ```
+        )",
+    });
 
     v.mkBool(true);
-    addConstant("true", v);
+    addConstant("true", v, {
+        .type = "boolean",
+        .doc = R"(
+          The boolean value for truth.
+
+          Users familiar with other languages might assume `true` is special syntax in the Nix language,
+          but this is in fact not the case:
+
+          ```nix-repl
+          nix-repl> let true = 1; in true
+          1
+          ```
+        )",
+    });
 
     v.mkBool(false);
-    addConstant("false", v);
+    addConstant("false", v, {
+        .type = "boolean",
+        .doc = R"(
+          The boolean value for falsity.
+
+          Users familiar with other languages might assume `false` is special syntax in the Nix language,
+          but this is in fact not the case:
+
+          ```nix-repl
+          nix-repl> let false = 1; in false
+          1
+          ```
+        )",
+    });
 
     v.mkNull();
-    addConstant("null", v);
+    addConstant("null", v, {
+        .type = "null",
+        .doc = R"(
+          The null value.
+
+          Users familiar with other languages might assume `null` is special syntax in the Nix language,
+          but this is in fact not the case:
+
+          ```nix-repl
+          nix-repl> let null = 1; in null
+          1
+          ```
+        )",
+    });
 
     if (!evalSettings.pureEval) {
         v.mkInt(time(0));
-        addConstant("__currentTime", v);
+        addConstant("__currentTime", v, {
+            .type = "integer",
+            .doc = R"(
+              The current time.
+            )",
+        });
 
         v.mkString(settings.thisSystem.get());
-        addConstant("__currentSystem", v);
+        addConstant("__currentSystem", v, {
+            .type = "string",
+            .doc = R"(
+              The built-in value `currentSystem` evaluates to the Nix platform
+              identifier for the Nix installation on which the expression is being
+              evaluated, such as `"i686-linux"` or `"x86_64-darwin"`.
```

Comment:
This should link to the `system` setting, and it should be mentioned that this constant is not available in pure mode.
Here's something for primops instead:
 - https://github.com/NixOS/nix/pull/8371

Reply:
Why `system` setting, what does it have to do with it?

Reply:
It has the same value.

```
$ nix-instantiate --option system "foo bar" --eval --expr 'builtins.currentSystem'
"foo bar"
```

Considering this didn't happen
 - https://github.com/NixOS/nix/pull/4093

Reply:
Ah never mind, I confused the lines with `currentTime`

Reply:
Linked